--all-targets not always desirable

When building a project for a specific no-std target (in my case for `thumbv6m-none-eabi`), the argument `--all-targets` casues `cargo check` to fail because there is no crate `test`. The same thing happens when I build from the command line.

Please make it configurable or even just a tickbox or something! I've had to go back to an older version of the plugin before the `--all-targets` change was introduced and am stuck there.

An example log is shown:
```
C:/Users/David/.cargo/bin/cargo.exe build --all --all-targets
   Compiling coffee-compass v0.1.0 (file:///C:/Users/David/Documents/code/coffee-compass)
error[E0463]: can't find crate for `test`
 --> src\main.rs:1:1
  |
1 | #![no_std]
  | ^ can't find crate

error: aborting due to previous error
```

--- a/src/main/kotlin/org/rust/cargo/project/configurable/RustProjectConfigurable.kt
+++ b/src/main/kotlin/org/rust/cargo/project/configurable/RustProjectConfigurable.kt
@@ -32,26 +32,29 @@ class RustProjectConfigurable(
         project.cargoProjects.allProjects.firstOrNull()?.rootDir?.pathAsPath ?: Paths.get(".")
     )
 
-    private val autoUpdateEnabledCheckbox = JBCheckBox()
+    private val autoUpdateEnabledCheckbox: JBCheckBox = JBCheckBox()
     private var autoUpdateEnabled: Boolean by CheckboxDelegate(autoUpdateEnabledCheckbox)
 
-    private val useCargoCheckForBuildCheckbox = JBCheckBox()
+    private val useCargoCheckForBuildCheckbox: JBCheckBox = JBCheckBox()
     private var useCargoCheckForBuild: Boolean by CheckboxDelegate(useCargoCheckForBuildCheckbox)
 
-    private val useCargoCheckAnnotatorCheckbox = JBCheckBox()
+    private val useCargoCheckAnnotatorCheckbox: JBCheckBox = JBCheckBox()
     private var useCargoCheckAnnotator: Boolean by CheckboxDelegate(useCargoCheckAnnotatorCheckbox)
 
-    private val useOfflineForCargoCheckCheckbox = JBCheckBox()
+    private val useOfflineForCargoCheckCheckbox: JBCheckBox = JBCheckBox()
     private var useOfflineForCargoCheck: Boolean by CheckboxDelegate(useOfflineForCargoCheckCheckbox)
 
-    private val expandMacrosCheckbox = JBCheckBox()
+    private val compileAllTargetsCheckBox = JBCheckBox()
+    private var compileAllTargets: Boolean by CheckboxDelegate(compileAllTargetsCheckBox)
+
+    private val expandMacrosCheckbox: JBCheckBox = JBCheckBox()
     private var expandMacros: Boolean by CheckboxDelegate(expandMacrosCheckbox)
 
     private val hintProvider = InlayParameterHintsExtension.forLanguage(RsLanguage)
     private val hintCheckboxes: Map<String, JBCheckBox> =
         hintProvider.supportedOptions.associate { it.id to JBCheckBox() }
 
-    private fun checkboxForOption(opt: Option) = hintCheckboxes[opt.id]!!
+    private fun checkboxForOption(opt: Option): JBCheckBox = hintCheckboxes[opt.id]!!
 
     override fun createComponent(): JComponent = layout {
         rustProjectSettings.attachTo(this)
@@ -60,6 +63,7 @@ class RustProjectConfigurable(
             row("Watch Cargo.toml:", autoUpdateEnabledCheckbox)
             row("Use cargo check to analyze code:", useCargoCheckAnnotatorCheckbox)
             row("Use cargo check when build project:", useCargoCheckForBuildCheckbox)
+            row("Compile all project targets if possible:", compileAllTargetsCheckBox)
             row("Use '-Zoffline' for cargo check (nightly only):", useOfflineForCargoCheckCheckbox)
         }
         val supportedHintOptions = hintProvider.supportedOptions
@@ -85,6 +89,7 @@ class RustProjectConfigurable(
         autoUpdateEnabled = settings.autoUpdateEnabled
         useCargoCheckForBuild = settings.useCargoCheckForBuild
         useCargoCheckAnnotator = settings.useCargoCheckAnnotator
+        compileAllTargets = settings.compileAllTargets
         useOfflineForCargoCheck = settings.useOfflineForCargoCheck
         expandMacros = settings.expandMacros
 
@@ -108,6 +113,7 @@ class RustProjectConfigurable(
             autoUpdateEnabled = autoUpdateEnabled,
             useCargoCheckForBuild = useCargoCheckForBuild,
             useCargoCheckAnnotator = useCargoCheckAnnotator,
+            compileAllTargets = compileAllTargets,
             useOfflineForCargoCheck = useOfflineForCargoCheck,
             expandMacros = expandMacros
         )
@@ -122,6 +128,7 @@ class RustProjectConfigurable(
             || autoUpdateEnabled != settings.autoUpdateEnabled
             || useCargoCheckForBuild != settings.useCargoCheckForBuild
             || useCargoCheckAnnotator != settings.useCargoCheckAnnotator
+            || compileAllTargets != settings.compileAllTargets
             || useOfflineForCargoCheck != settings.useOfflineForCargoCheck
             || expandMacros != settings.expandMacros
     }

--- a/src/main/kotlin/org/rust/cargo/project/settings/RustProjectSettingsService.kt
+++ b/src/main/kotlin/org/rust/cargo/project/settings/RustProjectSettingsService.kt
@@ -20,6 +20,7 @@ interface RustProjectSettingsService {
         val explicitPathToStdlib: String?,
         val useCargoCheckForBuild: Boolean,
         val useCargoCheckAnnotator: Boolean,
+        val compileAllTargets: Boolean,
         val useOfflineForCargoCheck: Boolean,
         val expandMacros: Boolean
     )
@@ -35,6 +36,7 @@ interface RustProjectSettingsService {
     val autoUpdateEnabled: Boolean get() = data.autoUpdateEnabled
     val useCargoCheckForBuild: Boolean get() = data.useCargoCheckForBuild
     val useCargoCheckAnnotator: Boolean get() = data.useCargoCheckAnnotator
+    val compileAllTargets: Boolean get() = data.compileAllTargets
     val useOfflineForCargoCheck: Boolean get() = data.useOfflineForCargoCheck
     val expandMacros: Boolean get() = data.expandMacros
 

--- a/src/main/kotlin/org/rust/cargo/project/settings/impl/RustProjectSettingsServiceImpl.kt
+++ b/src/main/kotlin/org/rust/cargo/project/settings/impl/RustProjectSettingsServiceImpl.kt
@@ -28,6 +28,7 @@ class RustProjectSettingsServiceImpl(
         var explicitPathToStdlib: String? = null,
         var useCargoCheckForBuild: Boolean = true,
         var useCargoCheckAnnotator: Boolean = false,
+        var compileAllTargets: Boolean = true,
         var useOfflineForCargoCheck: Boolean = false,
         var expandMacros: Boolean = true
     )
@@ -51,6 +52,7 @@ class RustProjectSettingsServiceImpl(
                 explicitPathToStdlib = state.explicitPathToStdlib,
                 useCargoCheckForBuild = state.useCargoCheckForBuild,
                 useCargoCheckAnnotator = state.useCargoCheckAnnotator,
+                compileAllTargets = state.compileAllTargets,
                 useOfflineForCargoCheck = state.useOfflineForCargoCheck,
                 expandMacros = state.expandMacros
             )
@@ -62,6 +64,7 @@ class RustProjectSettingsServiceImpl(
                 explicitPathToStdlib = value.explicitPathToStdlib,
                 useCargoCheckForBuild = value.useCargoCheckForBuild,
                 useCargoCheckAnnotator = value.useCargoCheckAnnotator,
+                compileAllTargets = value.compileAllTargets,
                 useOfflineForCargoCheck = value.useOfflineForCargoCheck,
                 expandMacros = value.expandMacros
             )

--- a/src/main/kotlin/org/rust/cargo/runconfig/Utils.kt
+++ b/src/main/kotlin/org/rust/cargo/runconfig/Utils.kt
@@ -57,13 +57,16 @@ val Project.hasCargoProject: Boolean get() = cargoProjects.allProjects.isNotEmpt
 
 fun Project.buildProject() {
     val command = if (rustSettings.useCargoCheckForBuild) "check" else "build"
-    val allTargets = rustSettings.toolchain
-        ?.rawCargo()
-        ?.checkSupportForBuildCheckAllTargets() ?: false
+
     val arguments = mutableListOf("--all")
 
-    if (allTargets) {
-        arguments += "--all-targets"
+    if (rustSettings.compileAllTargets) {
+        val allTargets = rustSettings.toolchain
+            ?.rawCargo()
+            ?.checkSupportForBuildCheckAllTargets() ?: false
+        if (allTargets) {
+            arguments += "--all-targets"
+        }
     }
     if (rustSettings.useOfflineForCargoCheck) {
         arguments += "-Zoffline"

--- a/src/main/kotlin/org/rust/cargo/toolchain/Cargo.kt
+++ b/src/main/kotlin/org/rust/cargo/toolchain/Cargo.kt
@@ -15,6 +15,7 @@ import com.intellij.execution.process.ProcessOutput
 import com.intellij.openapi.Disposable
 import com.intellij.openapi.application.runReadAction
 import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.project.Project
 import com.intellij.openapi.util.Disposer
 import com.intellij.openapi.util.SystemInfo
 import com.intellij.openapi.vfs.VfsUtil
@@ -22,6 +23,7 @@ import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.util.net.HttpConfigurable
 import org.jetbrains.annotations.TestOnly
 import org.rust.cargo.CargoConstants.RUST_BACTRACE_ENV_VAR
+import org.rust.cargo.project.settings.rustSettings
 import org.rust.cargo.project.workspace.CargoWorkspace
 import org.rust.cargo.runconfig.runExecutable
 import org.rust.cargo.toolchain.impl.CargoMetadata
@@ -132,10 +134,10 @@ class Cargo(
     }
 
     @Throws(ExecutionException::class)
-    fun checkProject(owner: Disposable, projectDirectory: Path): ProcessOutput {
+    fun checkProject(project: Project, owner: Disposable, projectDirectory: Path): ProcessOutput {
         val arguments = mutableListOf("--message-format=json", "--all")
 
-        if (checkSupportForBuildCheckAllTargets()) {
+        if (project.rustSettings.compileAllTargets && checkSupportForBuildCheckAllTargets()) {
             arguments += "--all-targets"
         }
 

--- a/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
+++ b/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
@@ -126,7 +126,7 @@ private fun checkProject(info: CargoCheckAnnotationInfo): CargoCheckAnnotationRe
 
 
     val output = try {
-        info.toolchain.cargoOrWrapper(info.projectPath).checkProject(info.module, info.projectPath)
+        info.toolchain.cargoOrWrapper(info.projectPath).checkProject(info.module.project, info.module, info.projectPath)
     } catch (e: ExecutionException) {
         return null
     }

--- a/src/test/kotlin/org/rust/cargo/project/RustProjectSettingsServiceTest.kt
+++ b/src/test/kotlin/org/rust/cargo/project/RustProjectSettingsServiceTest.kt
@@ -8,6 +8,7 @@ package org.rust.cargo.project
 import com.intellij.configurationStore.deserialize
 import com.intellij.configurationStore.serialize
 import com.intellij.testFramework.LightPlatformTestCase
+import org.intellij.lang.annotations.Language
 import org.rust.cargo.project.settings.RustProjectSettingsService
 import org.rust.cargo.project.settings.impl.RustProjectSettingsServiceImpl
 import org.rust.cargo.toolchain.RustToolchain
@@ -19,12 +20,17 @@ import java.nio.file.Paths
 class RustProjectSettingsServiceTest : LightPlatformTestCase() {
     fun `test serialization`() {
         val service = RustProjectSettingsServiceImpl(LightPlatformTestCase.getProject())
+        @Language("XML")
         val text = """
             <State>
+              <option name="autoUpdateEnabled" value="false" />
+              <option name="compileAllTargets" value="false" />
+              <option name="expandMacros" value="false" />
               <option name="explicitPathToStdlib" value="/stdlib" />
               <option name="toolchainHomeDirectory" value="/" />
               <option name="useCargoCheckAnnotator" value="true" />
               <option name="useCargoCheckForBuild" value="false" />
+              <option name="useOfflineForCargoCheck" value="true" />
             </State>
         """.trimIndent()
         service.loadState(elementFromXmlString(text).deserialize())
@@ -35,12 +41,13 @@ class RustProjectSettingsServiceTest : LightPlatformTestCase() {
 
         check(service.data == RustProjectSettingsService.Data(
             toolchain = RustToolchain(Paths.get("/")),
-            autoUpdateEnabled = true,
+            autoUpdateEnabled = false,
             explicitPathToStdlib = "/stdlib",
             useCargoCheckForBuild = false,
             useCargoCheckAnnotator = true,
-            useOfflineForCargoCheck = false,
-            expandMacros = true
+            compileAllTargets = false,
+            useOfflineForCargoCheck = true,
+            expandMacros = false
         ))
     }
 }

